flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala

```diff
@@ -820,20 +820,59 @@ abstract class TableEnvironment(val config: TableConfig) {
 
   /**
     * Checks if a table is registered under the given name.
+    * Internal and external catalogs are both checked.
     *
     * @param name The table name to check.
     * @return true, if a table is registered under the name, false otherwise.
     */
   protected[flink] def isRegistered(name: String): Boolean = {
-    rootSchema.getTableNames.contains(name)
+    var isRegistered = rootSchema.getTableNames.contains(name)
+
+    // check if the table exists in external catalogs
+    if (!isRegistered) {
+      val (externalSchema, externalTableName) = resolveExternalTable(name)
+      if (externalSchema != null) {
+        isRegistered = externalSchema.getTableNames.contains(externalTableName)
+      }
+    }
+
+    return isRegistered
   }
 
   protected def getTable(name: String): org.apache.calcite.schema.Table = {
```

Comment:
the method should return `Option[org.apache.calcite.schema.Table]`. We need to adjust `StreamTableEnvironment` and `BatchTableEnvironment` for this and not wrap the result of the `getTable()` calls another time in `Option`.